[GPU] Add hard FPS limit without vsync

mdukat opened the pull request:
For some reason, games on my PC stutter when using vsync. Here's my proposition, for people who don't care about little tearing here or there.

My commit adds two new cvars:
- `hard_fps_enable` - Enables hard FPS limit (default false)
- `hard_fps` - Hard FPS value (default 60)

Hard FPS limit is ignored when `vsync` is enabled. Disabling hard FPS limit, makes emulator run like it was before, with 1ms delay in `GraphicsSystem::Setup`.

Successfully built and tested with `xb build --config=release`.

Pipi86 commented:
I think it would be more user friendly if you rename the cvars "hard_fps_enable" > "fps_limiter_enable", and "hard_fps" > "fps_limit".
But anyways thanks for adding this.

Gliniak commented:
Better idea!
Instead of creating another code for limiting framerate maybe it would be better to reuse (as much as possible) vsync one?
Instead of calling everything vsync_worker_thread_ etc just call it frame_limiter_worker_thread_ or something.
That way vsync_fps->framerate_limit.

Default value would be 0 if you want uncapped and any value above 0 for some cap

Scenarios:
Vsync enabled & value is set: Execute Vsync version
Vsync disabled & value is set: Execute normal limiter
Vsync enabled & value is not set: Use some default value
Vsync disabled & value is not set: Full throtle

mdukat commented:
> I think it would be more user friendly if you rename the cvars "hard_fps_enable" > "fps_limiter_enable", and "hard_fps" > "fps_limit".

That's what I was thinking. However, my code is not as smart, as I would like it to be, if named "fps_limit". I believe for that, it would have to measure the time it took to calculate and render the frame, and then sleep the rest of the time, to hit specified framerate. I tried to do that, but ultimately failed. Hence, "hard" in the name.

> But anyways thanks for adding this.

You're welcome!

> Instead of creating another code for limiting framerate maybe it would be better to reuse (as much as possible) vsync one?
> Instead of calling everything vsync_worker_thread_ etc just call it frame_limiter_worker_thread_ or something.
> That way vsync_fps->framerate_limit.
> 
> Default value would be 0 if you want uncapped and any value above 0 for some cap

Yep, sounds good. Believe I could do that. Will commit in this PR.

___

Also, for anyone also having problems with vsync, with nvidia GPU's: check your drivers version. In 551.52, nvidia fixed "intermittent micro-stuttering". Basically, I was fooled by the oldest joke in the book: outdated drivers. Will update today, and I'll let you know if it helps in the next comment, to keep the PR neat and clean.

![image](https://github.com/xenia-canary/xenia-canary/assets/27694513/9441a2c5-5b8a-43fc-9f80-e57b45573581)

https://us.download.nvidia.com/Windows/551.52/551.52-win11-win10-release-notes.pdf

mdukat commented:
Modified `vsync_worker_thread` as per @Gliniak comment. Renamed it to `frame_limiter_worker_thread`.

Tested with scenarios:
 - vsync false, framerate_limit 0: unlimited frames
 - vsync true, framerate_limit 0: VSYNC enabled, locked at (default) 60 fps
 - vsync false, framerate_limit 75: VSYNC disabled, frames locked at 75 fps
 - vsync true, framerate_limit 75: VSYNC enabled, locked at 75 fps
 
Further tests and comments are welcome.

___

And regarding nvidia drivers, update really helps.

Pipi86 commented:
I did some little testing with some games i have that are affected by disabling vsync (Halo 3, Dark Souls, Forza Horizon 1&2) and others that are not (Diablo III: RoS, Red Dead Redemption) and didn't noticed anything wrong.
One thing i want to note is that i had to use the search tool from Wordpad to find the "framerate_limit" setting in xenia cfg file, because i was expecting it to be somewhere around "vsync" setting.

mdukat committed:
[GPU] Change FPS limit strategy
